### ISSUE TITLE
Implement automatic checkpoint recovery for training

### ISSUE THREAD

**Issue by GarthDB** [GarthDB/metal-candle#47]
## Problem

Current implementation has no checkpoint saving/loading during training:
```rust
// From trainer.rs:470
// TODO: Implement checkpoint saving with metal-candle
return Err(MLXError::NotImplemented(
    "Checkpoint saving not yet implemented - needs metal-candle checkpoint API".to_string(),
));
```

**Consequences:**
- Long training runs lost if interrupted (power failure, crash, etc.)
- No way to resume training from intermediate state
- Difficult to track training progress across sessions
- Can't checkpoint and continue later

## Proposed Solution

Implement automatic checkpoint saving and recovery system for reliable long-running training.

### Core API

```rust
pub struct CheckpointConfig {
    pub enabled: bool,
    pub checkpoint_dir: PathBuf,
    pub save_interval: CheckpointInterval,
    pub keep_last_n: usize,           // Keep only N recent checkpoints
    pub save_optimizer_state: bool,   // Include optimizer in checkpoint
    pub compression: bool,            // Compress checkpoint files
}

pub enum CheckpointInterval {
    Steps(usize),              // Every N steps
    Epochs(usize),             // Every N epochs
    Time(Duration),            // Every N minutes
    Loss(f32),                 // When loss improves by threshold
    Auto,                      // Smart checkpointing
}

pub struct Checkpoint {
    pub step: usize,
    pub epoch: usize,
    pub model_state: ModelState,
    pub optimizer_state: Option<OptimizerState>,
    pub training_config: TrainingConfig,
    pub loss: f32,
    pub timestamp: DateTime<Utc>,
    pub metadata: HashMap<String, String>,
}
```

### Training API with Checkpointing

```rust
impl Trainer {
    /// Train with automatic checkpointing
    pub fn train_with_checkpoints(
        &mut self,
        dataset: &Dataset,
        config: TrainingConfig,
        checkpoint_config: CheckpointConfig,
    ) -> Result<TrainingResult>;
    
    /// Resume training from latest checkpoint
    pub fn resume_from_latest_checkpoint(
        checkpoint_dir: &Path,
    ) -> Result<Self>;
    
    /// Resume from specific checkpoint
    pub fn resume_from_checkpoint(
        checkpoint_path: &Path,
    ) -> Result<Self>;
    
    /// Save checkpoint manually
    pub fn save_checkpoint(&self, path: &Path) -> Result<()>;
    
    /// List available checkpoints
    pub fn list_checkpoints(checkpoint_dir: &Path) -> Result<Vec<CheckpointInfo>>;
    
    /// Clean old checkpoints (keep best/recent N)
    pub fn clean_checkpoints(
        checkpoint_dir: &Path,
        strategy: CleanupStrategy,
    ) -> Result<usize>;
}
```

### 1. Automatic Recovery from Failures
```rust
// Set up automatic checkpointing
let checkpoint_config = CheckpointConfig {
    enabled: true,
    checkpoint_dir: PathBuf::from("checkpoints/qwen-rust-style"),
    save_interval: CheckpointInterval::Steps(100),  // Every 100 steps
    keep_last_n: 5,  // Keep 5 most recent
    save_optimizer_state: true,
    compression: true,
};

// Start training (auto-saves every 100 steps)
let result = trainer.train_with_checkpoints(
    &dataset,
    training_config,
    checkpoint_config,
)?;

// If interrupted, resume automatically:
let trainer = Trainer::resume_from_latest_checkpoint(
    "checkpoints/qwen-rust-style"
)?;

trainer.train_with_checkpoints(&dataset, training_config, checkpoint_config)?;
// Continues from step 437 (where it left off)
```

### 2. Long-Running Training Sessions
```rust
// Train overnight with checkpoints every hour
let config = CheckpointConfig {
    save_interval: CheckpointInterval::Time(Duration::from_secs(3600)),
    checkpoint_dir: PathBuf::from("overnight-training"),
    keep_last_n: 10,  // Keep 10 hours of checkpoints
    // ...
};

// Start training, go to sleep 😴
trainer.train_with_checkpoints(&dataset, training_config, config)?;

// Next morning: analyze checkpoints
let checkpoints = Trainer::list_checkpoints("overnight-training")?;
for ckpt in checkpoints {
    println!("Step {}: Loss = {:.4}", ckpt.step, ckpt.loss);
}
```

### 3. Experiment with Different Hyperparameters
```rust
// Save checkpoint at epoch 5
trainer.save_checkpoint("checkpoint-epoch5.safetensors")?;

// Try different learning rates from same starting point
for lr in [1e-4, 5e-5, 1e-5] {
    let mut trainer = Trainer::resume_from_checkpoint(
        "checkpoint-epoch5.safetensors"
    )?;
    
    trainer.config.learning_rate = lr;
    let result = trainer.train(&dataset)?;
    
    results.insert(lr, result.final_loss);
}
```

### 4. Best Checkpoint Selection
```rust
// Train with validation-based checkpointing
let config = CheckpointConfig {
    save_interval: CheckpointInterval::Loss(0.01),  // Save when loss improves
    keep_last_n: 3,  // Keep 3 best checkpoints
    // ...
};

// After training, load best checkpoint
let checkpoints = Trainer::list_checkpoints("checkpoints")?;
let best = checkpoints.iter().min_by_key(|c| c.loss).unwrap();
let best_model = Trainer::resume_from_checkpoint(&best.path)?;
```

### Safetensors (Recommended)
```
checkpoint-step-1000.safetensors
├── model/lora_A.q_proj        [tensor]
├── model/lora_B.q_proj        [tensor]
├── model/lora_A.v_proj        [tensor]
├── model/lora_B.v_proj        [tensor]
├── optimizer/adam_m           [tensor]
├── optimizer/adam_v           [tensor]
└── metadata
    ├── step: 1000
    ├── epoch: 3
    ├── loss: 0.452
    ├── learning_rate: 1e-4
    └── timestamp: 2025-12-12T10:30:00Z
```

**Advantages:**
- Fast loading (mmap support)
- Corruption detection (checksums)
- Metadata storage
- Language-agnostic format

## Directory Structure

```
checkpoints/qwen-rust-style/
├── checkpoint-step-0100.safetensors    (auto-saved)
├── checkpoint-step-0200.safetensors    (auto-saved)
├── checkpoint-step-0300.safetensors    (auto-saved)
├── checkpoint-epoch-1.safetensors      (epoch boundary)
├── checkpoint-best-loss.safetensors    (best so far)
├── checkpoint-final.safetensors        (training complete)
└── training_metadata.json              (overall info)
```

## Smart Checkpointing Strategy

```rust
pub enum CheckpointInterval {
    Auto,  // Intelligent checkpointing
}

// Auto strategy:
// - Save every 100 steps for first 1000 steps (rapid changes)
// - Save every 500 steps for next 5000 steps
// - Save every 1000 steps after that
// - Always save at epoch boundaries
// - Always save when loss improves by >1%
// - Keep best 3 checkpoints by loss
// - Keep last 3 checkpoints by time
```

## Checkpoint Cleanup

```rust
pub enum CleanupStrategy {
    KeepLastN(usize),           // Keep N most recent
    KeepBestN(usize),           // Keep N with lowest loss
    KeepBoth { last: usize, best: usize },  // Keep both
    KeepInterval(Duration),     // Keep checkpoints older than N days
}

// Clean up old checkpoints
Trainer::clean_checkpoints(
    "checkpoints/qwen-rust-style",
    CleanupStrategy::KeepBoth { last: 3, best: 3 },
)?;
// Removed 47 old checkpoints, kept 6 (3 recent + 3 best)
```

## Recovery UI/CLI

```bash
# List checkpoints
$ ferris mlx checkpoints list checkpoints/qwen-rust-style
Step    Epoch   Loss     Time                 Size
100     0       0.823    2025-12-12 10:15:23  142MB
200     0       0.612    2025-12-12 10:18:45  142MB
300     1       0.487    2025-12-12 10:22:11  142MB
400     1       0.421    2025-12-12 10:25:37  142MB (best)
500     2       0.398    2025-12-12 10:29:02  142MB (best) (latest)

# Resume training
$ ferris mlx train --resume checkpoints/qwen-rust-style
✓ Resuming from checkpoint-step-0500.safetensors
✓ Step: 500, Epoch: 2, Loss: 0.398
⚡ Continuing training...

# Resume from specific checkpoint
$ ferris mlx train --checkpoint checkpoints/qwen-rust-style/checkpoint-step-0300.safetensors
✓ Resuming from step 300
```

## Technical Requirements

1. **Atomic Writes**: Prevent corruption from interrupted saves
2. **Validation**: Verify checkpoint integrity on load
3. **Compression**: Optional gzip/zstd compression
4. **Metadata**: Store config, step, epoch, loss, timestamp
5. **Backwards Compat**: Handle checkpoints from older versions
6. **Error Handling**: Graceful fallback if checkpoint corrupt

## Performance Impact

| Operation | Time | Impact |
|-----------|------|--------|
| Save checkpoint (7B LoRA) | 1-2s | Low |
| Load checkpoint | 0.5-1s | Low |
| Training with checkpointing | +1% overhead | Minimal |
| Checkpoint compression | +50% save time | Optional |

**Recommendation**: Save every 100-500 steps (overhead: 0.2-1% of training time)

## Benefits for Ferris

- **Training reliability** (never lose progress)
- **Experiment faster** (checkpoint and try variations)
- **Resume interrupted runs** (power failure, crash, etc.)
- **Track progress** (see loss curve over time)
- **Select best model** (not just final model)
- **Peace of mind** (long runs won't be lost)

## Priority

**MEDIUM** - Important for reliability but not blocking current workflows

## Related Issues

- Complements #46 (Mixed Precision) for faster checkpoints
- Useful for long runs with #44 (Extended Context)
- Important for #45 (Multi-Adapter) management

## Prior Art

- PyTorch checkpointing: https://pytorch.org/tutorials/recipes/recipes/saving_and_loading_checkpoints_for_inference.html
- HuggingFace Trainer: https://huggingface.co/docs/transformers/main_classes/trainer#checkpointing
- Safetensors format: https://github.com/huggingface/safetensors

### Phase 1: Basic Checkpointing
- ✅ Save/load model state (LoRA weights)
- ✅ Save/load optimizer state
- ✅ Metadata storage
- ✅ Manual checkpoint API

### Phase 2: Automatic Checkpointing
- ✅ Interval-based saving (steps/epochs)
- ✅ Resume from latest
- ✅ Training integration

### Phase 3: Advanced Features
- Smart checkpointing strategy
- Checkpoint cleanup
- Best checkpoint selection
- CLI commands
